modules/3d/test/test_tsdf.cpp

```diff
@@ -1585,36 +1421,75 @@ TEST(ColorTSDF_CPU, valid_points_common_framesize_fetch)
     cv::ocl::setUseOpenCL(true);
 }
 
-TEST(ColorTSDF_CPU, boundingBox)
-{
-    cv::ocl::setUseOpenCL(false);
-    boundingBoxGrowthTest(VolumeType::ColorTSDF);
-    cv::ocl::setUseOpenCL(true);
-}
 
-TEST(HashTSDF_CPU, boundingBoxEnableGrowth)
+class StaticVolumeBoundingBox : public ::testing::TestWithParam<VolumeType>
+{ };
+
+TEST_P(StaticVolumeBoundingBox, boundingBoxCPU)
 {
-    cv::ocl::setUseOpenCL(false);
-    boundingBoxGrowthTest(VolumeType::HashTSDF);
-    cv::ocl::setUseOpenCL(true);
+    bool revertOcl = cv::ocl::useOpenCL();
+    if (revertOcl)
+        cv::ocl::setUseOpenCL(false);
+
+    staticBoundingBoxTest(GetParam());
+
+    if (revertOcl)
+        cv::ocl::setUseOpenCL(true);
 }
 
+INSTANTIATE_TEST_CASE_P(TSDF, StaticVolumeBoundingBox, ::testing::Values(VolumeType::TSDF, VolumeType::ColorTSDF));
+
+
 // OpenCL tests
 TEST(HashTSDF_GPU, reproduce_volPoseRot)
 {
     regressionVolPoseRot();
 }
 
-TEST(TSDF_GPU, boundingBox)
+//TODO: use this when ColorTSDF gets GPU support
+/*
+class StaticVolumeBoundingBox : public ::testing::TestWithParam<VolumeType>
+{ };
+
+TEST_P(StaticVolumeBoundingBox, GPU)
 {
-    boundingBoxGrowthTest(VolumeType::TSDF);
+    staticBoundingBoxTest(GetParam());
 }
 
-TEST(HashTSDF_GPU, boundingBoxEnableGrowth)
+INSTANTIATE_TEST_CASE_P(TSDF, StaticVolumeBoundingBox, ::testing::Values(VolumeType::TSDF, VolumeType::ColorTSDF));
+*/
+
+// until that, we don't need parametrized test
+TEST(TSDF, boundingBoxGPU)
 {
-    boundingBoxGrowthTest(VolumeType::HashTSDF);
+    staticBoundingBoxTest(VolumeType::TSDF);
 }
 
+
+class BoundingBoxEnableGrowthTest : public ::testing::TestWithParam<std::tuple<bool, bool>>
+{ };
+
+TEST_P(BoundingBoxEnableGrowthTest, boundingBoxEnableGrowth)
+{
+    auto p = GetParam();
+    bool gpu = std::get<0>(p);
+    bool enableGrowth = std::get<1>(p);
+
+    bool revertOcl = false;
+    if (!gpu)
+        revertOcl = cv::ocl::useOpenCL();
+
+    if (revertOcl)
+        cv::ocl::setUseOpenCL(false);
+
+    boundingBoxGrowthTest(enableGrowth);
```

Comment:
In case if `boundingBoxGrowthTest` throws exception or fails the test the next step for OpenCL will not be executed. I propose to use special struct with constructor+destructor to guarantee context handling. It's applicable for other OpenCL specific tests too:

```
struct OpenCLContextStorage
{
     OpenCLContextStorage(bool openclStatus):
        originalOpenCLStatus(openclStatus)
    {}
     ~OpenCLContextStorage()
    {
        cv::ocl::setUseOpenCL(originalOpenCLStatus);
    }
}
```

And in test:
```
    OpenCLContextStorage oclStatusStorage(false);
   ...
```

Reply:
Done
As for other tests, this will be done in the next PR